### PR TITLE
Work branch

### PR DESCRIPTION
**Pull request by aburluka**

### CODE REVIEW

**Review by dmission — APPROVED** [review]

File: vzstorage-pd/vendor/github.com/virtuozzo/goploop-cli/ploopvol/ploop-volume.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+	"os/exec"
+	"path"
+)
+
+var (
+	ploopVolumeBinary = "/usr/sbin/ploop-volume"
+)
+
+type PloopVolume struct {
+	Path string
```
Comment (avagin) [inline vzstorage-pd/vendor/github.com/virtuozzo/goploop-cli/ploopvol/ploop-volume.go]: Who has to set Path? Can we have something like func Open(volumePath string) (PloopVolume, error)?

File: vzstorage-pd/vendor/github.com/virtuozzo/goploop-cli/ploopvol/ploop-volume.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+func (pv *PloopVolume) Create(size uint64, image string) error {
```
Comment (avagin) [inline vzstorage-pd/vendor/github.com/virtuozzo/goploop-cli/ploopvol/ploop-volume.go]: I think it is better to use the same interface with ploop.Create(). The library has to follow the same principals for all functions.

File: vzstorage-pd/vendor/github.com/virtuozzo/goploop-cli/ploopvol/ploop-volume.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+			return fmt.Errorf("Cannot create backup of destination %s: %v", dst, err)
+		}
+	}
+
+	args := []string{"clone", dst, pv.Path}
+	cmd := exec.Command(ploopVolumeBinary, args...)
+	if err := cmd.Run(); err != nil {
+		if _, err := os.Stat(backupPath); !os.IsNotExist(err) {
+			if err = os.Rename(backupPath, pv.Path); err != nil {
+				return fmt.Errorf("Cannot restore backup %s: %v", pv.Path, err)
```
Comment (avagin) [inline vzstorage-pd/vendor/github.com/virtuozzo/goploop-cli/ploopvol/ploop-volume.go]: All functions has to return ploop.Err

File: vzstorage-pd/vendor/github.com/virtuozzo/goploop-cli/ploopvol/ploop-volume.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+func (pv *PloopVolume) Clone(dst string) error {
+	if dst == "" {
+		return fmt.Errorf("Bad destination path!")
+	}
+	backupPath := path.Clean(dst) + ".backup"
+
+	if _, err := os.Stat(pv.Path); !os.IsNotExist(err) {
+		if err = os.Rename(pv.Path, backupPath); err != nil {
```
Comment (avagin) [inline vzstorage-pd/vendor/github.com/virtuozzo/goploop-cli/ploopvol/ploop-volume.go]: I don't like this game with backupPath(). Clone() has to fail if a destination path already exists.

File: vzstorage-pd/vendor/github.com/virtuozzo/goploop-cli/ploopvol/ploop-volume.go (hunk 1 above)
Comment (aburluka) [reply vzstorage-pd/vendor/github.com/virtuozzo/goploop-cli/ploopvol/ploop-volume.go]: This abstraction makes additional mess for me without any significant benefit

File: vzstorage-pd/vendor/github.com/virtuozzo/goploop-cli/ploopvol/ploop-volume.go (hunk 4 above)
Comment (aburluka) [reply vzstorage-pd/vendor/github.com/virtuozzo/goploop-cli/ploopvol/ploop-volume.go]: Igor introduces ploop-volume switch routine so I will rework that part

File: vzstorage-pd/vendor/github.com/virtuozzo/goploop-cli/ploop-volume.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+func PloopVolumeOpen(src string) (*PloopVolume, error) {
+	if _, err := os.Stat(path.Join(src, "DiskDescriptor.xml")); os.IsNotExist(err) {
+		return nil, &Err{c:-1, s: fmt.Sprintf("Bad ploop-volume path %s!", src)}
+	}
+	return &PloopVolume{src}, nil
+}
+
+func PloopVolumeCreate(src string, size uint64, image string) (*PloopVolume, error) {
```
Comment (avagin) [inline vzstorage-pd/vendor/github.com/virtuozzo/goploop-cli/ploop-volume.go]: Maybe we can use CreateParam here like ploop.Create(p *CreateParam)

File: vzstorage-pd/vendor/github.com/virtuozzo/goploop-cli/ploop-volume.go [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+		return &Err{c:-1, s: fmt.Sprintf("Bad destination path!")}
+	}
+	return ploopVolume("snapshot", pv.Path, dst)
+}
+
+func (pv *PloopVolume) Switch(dst string) error {
+	if dst == "" {
+		return &Err{c:-1, s: fmt.Sprintf("Bad destination path!")}
+	}
+	if _, err := os.Stat(pv.Path); os.IsNotExist(err) {
```
Comment (avagin) [inline vzstorage-pd/vendor/github.com/virtuozzo/goploop-cli/ploop-volume.go]: Why do we need this check?

File: vzstorage-pd/vendor/github.com/virtuozzo/goploop-cli/ploop-volume.go [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+func (pv *PloopVolume) Clone(dst string) error {
+	if dst == "" {
+		return &Err{c:-1, s: fmt.Sprintf("Bad destination path!")}
+	}
+
+	if _, err := os.Stat(pv.Path); !os.IsNotExist(err) {
+		return &Err{c:-1, s: fmt.Sprintf("Clone destination is already exist %s: %v", dst, err)}
+	}
+
+	return ploopVolume("clone", dst, pv.Path)
```
Comment (avagin) [inline vzstorage-pd/vendor/github.com/virtuozzo/goploop-cli/ploop-volume.go]: Here is something wrong. 

[root@localhost ploop]# ./tools/ploop-volume clone 
Usage: ploop-volume clone <SRC> <DST>

Maybe you wanted to say this:
return ploopVolume("clone", pv.Path, dst)

File: vzstorage-pd/vendor/github.com/virtuozzo/goploop-cli/ploop-volume.go [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+	return ploopVolume("switch", pv.Path, dst)
+}
+
+func (pv *PloopVolume) Clone(dst string) error {
+	if dst == "" {
+		return &Err{c:-1, s: fmt.Sprintf("Bad destination path!")}
+	}
+
+	if _, err := os.Stat(pv.Path); !os.IsNotExist(err) {
+		return &Err{c:-1, s: fmt.Sprintf("Clone destination is already exist %s: %v", dst, err)}
```
Comment (avagin) [inline vzstorage-pd/vendor/github.com/virtuozzo/goploop-cli/ploop-volume.go]: You check pv.Path, but the error message about dst

File: vzstorage-pd/vendor/github.com/virtuozzo/goploop-cli/ploop-volume.go (hunk 5 above)
Comment (aburluka) [reply vzstorage-pd/vendor/github.com/virtuozzo/goploop-cli/ploop-volume.go]: It is not fully supported, so I'd rather not use it

File: vzstorage-pd/vendor/github.com/virtuozzo/goploop-cli/ploop-volume.go (hunk 6 above)
Comment (aburluka) [reply vzstorage-pd/vendor/github.com/virtuozzo/goploop-cli/ploop-volume.go]: Actually we can deal without it, but it makes this function a little bit more robust. Depends on you, I can easily remove it

File: vzstorage-pd/virtuozzo-provisioner.go [hunk 9]
```diff
@@ -162,12 +162,8 @@ func createPloop(mount string, options map[string]string) error {
 	}
 
 	// Create the ploop volume
-	cp := ploop.CreateParam{Size: volumeSize, File: ploopPath + "/" + options["volumeID"]}
-	if err := ploop.Create(&cp); err != nil {
-		return err
-	}
-
-	return nil
+	_, err = ploop.PloopVolumeCreate(ploopPath, volumeSize, "")
```
Comment (avagin) [inline vzstorage-pd/virtuozzo-provisioner.go]: Here is volumeSize in Kilobytes. Is it ok?

Comment (avagin) [reply vzstorage-pd/virtuozzo-provisioner.go]: I don't understand why the third argument is an empty string. I think we have to add a new parameter to a storage class which will say where delta-s have to be created.  We can call it as "deltaPath". If deltaPath isn't set we probably should create a regular ploop.

File: vzstorage-pd/vendor/github.com/virtuozzo/goploop-cli/ploop_volume.go [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+func PloopVolumeSnapshotOpen(src string) (*PloopVolumeSnapshot, error) {
+	if err := checkDD(src); err != nil {
+		return nil, err
+	}
+	return &PloopVolumeSnapshot{src}, nil
+}
+
+func PloopVolumeCreate(src string, size uint64, image string) (*PloopVolume, error) {
```
Comment (avagin) [inline vzstorage-pd/vendor/github.com/virtuozzo/goploop-cli/ploop_volume.go:43]: nit: s/src/path/

Comment (avagin) [reply vzstorage-pd/vendor/github.com/virtuozzo/goploop-cli/ploop_volume.go:43]: ploop.Create gets size in Kilobytes. Should we do here the same?

File: vzstorage-pd/vendor/github.com/virtuozzo/goploop-cli/ploop_volume.go [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+	}
+	return ploopVolume("switch", pvs.Path, pv.Path)
+}
+
+func (pvs *PloopVolumeSnapshot) Clone(dst string) (*PloopVolume, error) {
+	if dst == "" {
+		return nil, &Err{c:-1, s: fmt.Sprintf("Bad destination path!")}
+	}
+
+	if _, err := os.Stat(dst); os.IsNotExist(err) {
```
Comment (avagin) [inline vzstorage-pd/vendor/github.com/virtuozzo/goploop-cli/ploop_volume.go]: dst is a path to a new volume, why should it exist?

**Pull request merged**